### ISSUE TITLE
document development / branching model

### ISSUE THREAD

**Issue by maxheld83** [aiorazabala/qmethod#38]
We don't have a terrible lot of development (let alone developers) in here, but maybe some thoughts on a development model would be in order.

I'd suggest the following steps:

1. Log issues for todos
  -  assign `backlog` milestone if it's not happening anytime soon
  -  assign oneself as an assignee if one is actively working on it (so as to avoid duplicate efforts)
2. Collaborators: Create forks
  - such that other collaborators can work (and mess up) in their own sandbox
  - but, remember to pull in upstream changes from `aiorazabala/qmethod/master` frequently, so as to stay up to date and avoid merge conflicts.
3. For bigger things, create "feature-branches"
  - create a branch off of master for some feature to be added, say `rotation-visualization` (hehe, I wish)
  - git afficionados seem to urge to do this even for small stuff
4. Once work is done, put up a pull-request (see #36), which @aiorazabala as the maintainer and creator then accepts after review.
  - Pull requests should be put up to the `dev` branch.
  - **Crucially** pull requests should be put up only once all `R CMD check` passes.
  - @aiorazabala accepts the pull request once the `feature-branch` is ready for prime-time, then merged into `dev`
5. Periodically, whenever significant work has been done, @aiorazabala drafts a release from `master`, essentially just marking some point in the history of the package as `x.x.x`, and sends it off to CRAN.
  - very importantly, @aiorazabala should not have to squash any `R CMD check` messages or write docs. Pull requesters need to do this themselves.

**Comment by maxheld83** [comment]
If we agree on this, I'll be happy to append this to the `README.md` as a how-2-contribute section.

**Comment by aiorazabala** [comment]
Sounds good! feel free to add it to README, I'd add this as 'suggestions' or 'good practice', rather than compulsory type of thing (to not pull anyone off)

**Comment by maxheld83** [comment]
@aiorazabala absolutely! 
... and it's not like we're drowning in contributions :wink: 

(I'll be using a lot more of [these](http://www.emoji-cheat-sheet.com), I think. Add some color to my grey office.

**Comment by maxheld83** [comment]
done!
Hope I got the tone right.

**Comment by maxheld83** [comment]
uh just FYI, *[this](https://guides.github.com/introduction/flow/)* is basically what we're doing now.

**Comment by maxheld83** [comment]
since some discussion on this came up, I'll open this again.

**Comment by maxheld83** [comment]
@aiorazabala can you chime in on this again? Does [this](https://guides.github.com/introduction/flow/index.html) (see above) make sense for you?

I'm still wondering whether it might not make more sense to point (my) pull requests to your `master`, because I think these pull requests really *should* remain open, unmerged until they are CRAN-ready.
And once they are, they might as well go to `master`, or?
That might make things a little easier, especially by *not* polluting your `dev` branch, on which, I understand, you're doing your work.

Might be cleaner.

What do you think?

**Comment by aiorazabala** [comment]
As I understand it, the master is the branch where clean and most tested code is, and dev is the branch were new stuff goes. Thus first to dev, then to master, in my view.

**Comment by maxheld83** [comment]
mmh, I'm not sure – maybe we could talk about the whole collaboration model on Skype sometime soon, if you have the time, @aiorazabala?

I would maybe advocate for a [feature branching model](https://guides.github.com/introduction/flow/), where I develop new functions (or bug fixes) in branches (off `master`, respectively), and once done, put up a pull request back to `master`.
That way, each function or bug fix lives in it's own branch, and work on it does not pollute / break other stuff.

I agree that anything on `master` must be absolutely clean and tested, ready to be deployed to CRAN, but I would think that the Pull Request discussions and review take care of that, where some branch (say, `array-visualization`), *must* remain unmerged and open until and unless it's ready for CRAN.

Merging a feature branch into `dev` might muddy things a little, because it might pollute other work on that branch, and also creates a need for even more merges.

Anyway, I'd live to understand your thinking about this and how you work best – maybe this can best be done in a call.

**Comment by maxheld83** [comment]
maybe something like this?


![img_0052](https://cloud.githubusercontent.com/assets/5372770/6410913/8c7cbac4-be73-11e4-81ed-146569f71b02.JPG)

**Comment by maxheld83** [comment]
we decided to kill `dev`, and only work with feature branches.

Code review may for now proceed by downloading `*.zip`s.

In future, we'll try and use more of git(hub)s functionality.

Here, again, is how to clone pull request branches locally: https://help.github.com/articles/checking-out-pull-requests-locally/
